integration/spark-common/src/main/java/org/apache/carbondata/spark/merger/CompactionResultSortProcessor.java

```diff
@@ -0,0 +1,407 @@
+    switch (aggType) {
+      case CarbonCommonConstants.BIG_DECIMAL_MEASURE:
+        value = ((org.apache.spark.sql.types.Decimal) value).toJavaBigDecimal();
+        return value;
+      default:
+        return value;
+    }
+  }
+
+  /**
+   * This method will read sort temp files, perform merge sort and add it to store for data loading
+   */
+  private void readAndLoadDataFromSortTempFiles() throws Exception {
+    try {
+      finalMerger.startFinalMerge();
+      while (finalMerger.hasNext()) {
+        Object[] rowRead = finalMerger.next();
+        CarbonRow row = new CarbonRow(rowRead);
+        // convert the row from surrogate key to MDKey
+        Object[] outputRow = CarbonDataProcessorUtil
+            .convertToMDKeyAndFillRow(row, segmentProperties, measureCount, noDictionaryCount,
+                segmentProperties.getComplexDimensions().size());
+        dataHandler.addDataToStore(outputRow);
+      }
+      dataHandler.finish();
+    } catch (CarbonDataWriterException e) {
+      LOGGER.error(e);
+      throw new Exception("Problem loading data during compaction: " + e.getMessage());
+    } catch (Exception e) {
+      LOGGER.error(e);
+      throw new Exception("Problem loading data during compaction: " + e.getMessage());
+    } finally {
+      if (null != dataHandler) {
+        try {
+          dataHandler.closeHandler();
+        } catch (CarbonDataWriterException e) {
+          LOGGER.error(e);
+          throw new Exception("Problem loading data during compaction: " + e.getMessage());
+        }
+      }
+    }
+  }
+
+  /**
+   * add row to a temp array which will we written to a sort temp file after sorting
+   *
+   * @param row
+   */
+  private void addRowForSorting(Object[] row) throws Exception {
+    try {
+      // prepare row array using RemoveDictionaryUtil class
+      sortDataRows.addRow(row);
+    } catch (CarbonSortKeyAndGroupByException e) {
+      LOGGER.error(e);
+      throw new Exception("Row addition for sorting failed during compaction: " + e.getMessage());
+    }
+  }
+
+  /**
+   * create an instance of sort data rows
+   */
+  private void initSortDataRows() throws Exception {
+    measureCount = carbonTable.getMeasureByTableName(tableName).size();
+    List<CarbonDimension> dimensions = carbonTable.getDimensionByTableName(tableName);
+    noDictionaryColMapping = new boolean[dimensions.size()];
+    int i = 0;
+    for (CarbonDimension dimension : dimensions) {
+      if (CarbonUtil.hasEncoding(dimension.getEncoder(), Encoding.DICTIONARY)) {
+        i++;
+        continue;
+      }
+      noDictionaryColMapping[i++] = true;
+      noDictionaryCount++;
+    }
+    dimensionColumnCount = dimensions.size();
+    SortParameters parameters = createSortParameters();
+    SortIntermediateFileMerger intermediateFileMerger = new SortIntermediateFileMerger(parameters);
+    this.sortDataRows = new SortDataRows(parameters, intermediateFileMerger);
```

Comment:
Now it is only supported onheap merge, but we can have unsafe merge as well by using UnsafeSortDataRows.
Any way you can put as TODO and will handle later